Implemented Switchable leg orientation

Maaz0070 opened the pull request.

Maaz0070 committed:
fixed error in base class rover_system

Maaz0070 committed:
merged changes with master

Maaz0070 committed:
changes to RoverSystem base class

Maaz0070 committed:
changes to RoverSystem base class

Maaz0070 committed:
merging with unit testing commit

Maaz0070 committed:
removed common mc data to base class

Maaz0070 committed:
removed common mc data to base class

Maaz0070 committed:
merged with master

Maaz0070 committed:
Switchable leg orientation implmeented

naterpotatoers commented on Drive/rover_drive_system.hpp:
```diff
@@ -408,11 +420,13 @@ class RoverDriveSystem : public sjsu::common::RoverSystem
   Wheel & left_wheel_;
   Wheel & right_wheel_;
   Wheel & back_wheel_;
+  Wheel arrays[3] = {left_wheel_, right_wheel_, back_wheel_};
```
Was this utilized anywhere?

Maaz0070 replied:
No. I Forgot to remove it.

naterpotatoers reviewed (changes requested):
@kammce Do you have any ideas for resolving SJSURoboticsTeam/urc-central-2021/issues/167

Maaz0070 committed:
moved library files into common folder

Maaz0070 committed:
made sure wheels are not moving before swithing orientation

Maaz0070 committed:
removed testing line in main file

naterpotatoers commented:
Which files did you directly modify from Dev2? We only need the files you modified. Everything else can be imported from dev2 still

Maaz0070 commented:
> 
Since the files I modified included other files, mainly in the util folder, I just added the util folder. Everything outside the util folder are files I modified. 
Ill change it to where we don't import the entire util folder and only the ones we need.

Maaz0070 committed:
removed util folder from Common

naterpotatoers commented on Drive/rover_drive_system.hpp:
```diff
@@ -43,10 +43,28 @@ class RoverDriveSystem : public sjsu::common::RoverSystem
   };
 
+  // example Call: drive.switchLegOrientation(drive, back_wheel, right_wheel, back_wheel)
+  void switchLegOrientation(RoverDriveSystem &drive, Wheel &left, Wheel &right, Wheel &back)
```
Possible Wheel Orientations
Left, Right, Back --> Back, Left, Right --> Right, Back, Left
Back, Left, Right 
Right, Back, Left  
Not Possible Wheel Orientations (inverts drive steer logic)
Left, Back, Right
Right, Right, Right
```suggestion
  void SwitchLegOrientation(RoverDriveSystem &drive, Wheel &left, Wheel &right, Wheel &back)
```

naterpotatoers commented:
As per the Discord thread 'Updating Libraries' we can try inheriting the RMD, Gpio, and CAN drivers and adding in our logic changes that way. Also good idea to just rename these drivers to something different

Maaz0070 committed:
edited leg orientation using arrays

Maaz0070 committed:
merged with master

Maaz0070 committed:
removed testing line in drive system

naterpotatoers commented on Drive/rover_drive_system.hpp:
```diff
@@ -41,20 +42,43 @@ class RoverDriveSystem : public sjsu::common::RoverSystem
+    [1] = right back left
+    [2] = back left right
+    */
+    if(left_wheel_->GetHubSpeed() == 0 && right_wheel_->GetHubSpeed() == 0 && back_wheel_->GetHubSpeed() == 0)
```
```suggestion
    if(IsStopped())
```
We have a helper function for this call

naterpotatoers commented on Drive/rover_drive_system.hpp:
```diff
@@ -41,20 +42,43 @@ class RoverDriveSystem : public sjsu::common::RoverSystem
+        right_wheel_ = wheels[(position + 1) % 3];
+        back_wheel_ = wheels[(position + 2) % 3];
+    }
+    else
+    {
+      sjsu::LogError("Cannot change leg orientation while wheels are moving!");
+    }
```
This should never be ran. Our code should be smart enough to stop itself before this function ever gets called imo

naterpotatoers commented on Drive/rover_drive_system.hpp:
```diff
@@ -41,20 +42,43 @@ class RoverDriveSystem : public sjsu::common::RoverSystem
   }
 
+  // example Call: drive.switchLegOrientation(drive, back_wheel, right_wheel, back_wheel)
+  void switchLegOrientation(int position)
```
```suggestion
  void SwitchLegOrientation(int position)
```

naterpotatoers reviewed (changes requested):
Make sure to fix all the unit tests for drive system so that they all pass

Maaz0070 committed:
used IsStopped in SwitchLegOrientation method

Maaz0070 committed:
Merge branch 'master' of https://github.com/SJSURoboticsTeam/urc-control_systems-2021 into Maaz-LegOrientation

Maaz0070 committed:
fixed test case for leg orientation

Maaz0070 committed:
Merge branch 'master' of https://github.com/SJSURoboticsTeam/urc-control_systems-2021 into Maaz-LegOrientation

Maaz0070 committed:
edited testing function

FluffyFTW committed:
changed public member variables to private

FluffyFTW committed:
Revert "changed public member variables to private"

This reverts commit 5d33258beaad9ea698d00888fc6f28c1d3360d7a.

FluffyFTW committed:
changed all wheels to fit the struct and added a getter for the wheels

FluffyFTW committed:
stashing local changes before revert

FluffyFTW committed:
Revert "Revert "changed public member variables to private""

This reverts commit c3a19c12d5fb81ae53b498948ce73741a4f98c7a.

FluffyFTW committed:
changing the variables back to public

naterpotatoers commented:
Is the PR good to close since other PR got merged with these changes?

Maaz0070 commented:
> Is the PR good to close since other PR got merged with these changes?

Yes
